File: dynamic_language_python/sudoku/sudoku.py
import numpy as np

from .grid import Grid

'''数独'''
class Sudoku(Grid):
    def __init__(self):
        super().__init__(9, 3)

    '''
    解析输入字符串
    '''
    def parse(self, input_string: str) -> bool:
        # 1. Check if input string size equals to the grid
        if len(input_string) != self._grid_size * self._grid_size:
            self._resetData()
            return False
        
        # 2. Check if the char is a number
        for i, number in enumerate(input_string):
            if int(number) < 0 or int(number) > self._grid_size:
                self._resetData()
                return False
            
            self._data[i // self._grid_size][i % self._grid_size] = int(number)
        
        # 3. Check if the sudoku is valid
        if not self.__isSudokuValid():
            self._resetData()
            return False
        
        return True

    '''
    推所有可能值
    '''
    def getInference(self, row, column) -> list:
        # 1. Find out all existed numbers in same row, column and box
        existed_numbers = set(self.getRow(row))
        existed_numbers.update(self.getColumn(column))
        existed_numbers.update(self.getBox(row, column))

        # 2. Inference not showed up numbers(diff of two sets)
        return list(set(range(1, self._grid_size + 1)) - existed_numbers)
    
    '''
    检查数组是否合法(同行，列，格中是否存在相同数字)
    '''
    def __isSudokuValid(self) -> bool:
        # arr里存在同样的数字
        def containSameNumber(arr: np.array):
            numbers_count = np.bincount(arr)
            return np.any(numbers_count[1:] > 1) # Remove number 0

        # 1. Check all rows
        for r in range(self._grid_size):
            if containSameNumber(self.getRow(r)):
                return False
    
        # 2. Check all columns
        for c in range(self._grid_size):
            if containSameNumber(self.getColumn(c)):
                return False
        
        # 3. Check all boxes
        for r in range(0, self._grid_size, self._box_size):
            for c in range(0, self._grid_size, self._box_size):
                if containSameNumber(self.getBox(r, c)):
                    return False
        
        return True

File: static_language_c_plus_plus/src/sudoku/sudoku.cpp
#include <set>
#include "sudoku.h"

Sudoku::Sudoku()
    : Grid(9, 3)
{
}

bool Sudoku::parse(std::string_view str)
{
    // 1. Check if its size equals to the size of the grid
    if (str.size() != _grid_size * _grid_size)
    {
        resetData();
        return false;
    }

    // Limit number between 0 and _grid_size
    auto isCharAValidNumber = [this](char c){
        return c >= '0' && c <= '0' + _grid_size;
    };

    // 2. Parse input string
    for (int i = 0; i < str.size(); ++i)
    {
        if (!isCharAValidNumber(str[i])) // Between 0 and _grid_size
        {
            resetData();
            return false;
        }
        _data[i / _grid_size][i % _grid_size] = str[i] - '0';
    }

    // 3. Check if the sudoku is valid
    if (!isAValidSudoku())
    {
        resetData();
        return false;
    }

    return true;
}

std::vector<int> Sudoku::getInference(int row, int column) const
{
    // 1. Find out all numbers in the same row, column and box
    std::set<int> existed_numbers;
    // Add data to set
    auto addToSet = [&existed_numbers](const std::vector<int>& data) {
        existed_numbers.insert(data.begin(), data.end());
    };
    addToSet(getRow(row));          // Add same row numbers
    addToSet(getColumn(column));    // Add same column numbers
    addToSet(getBox(row, column));  // Add same box numbers
    
    // 2. Find out numbers not show yet
    std::vector<int> numbers_not_showed;
    for (int i = 1; i <= _grid_size; ++i)
    {
        if (existed_numbers.find(i) != existed_numbers.end())
            continue;

        numbers_not_showed.emplace_back(i);
    }
    return numbers_not_showed;
}

bool Sudoku::isAValidSudoku() const
{
    auto sameNumberExistsInOneArray = [this](const std::vector<int>& arr){
        std::set<int> unique_numbers;
        for(int i : arr)
        {
            // 1. Skip 0
            if (i == 0)
            {
                continue;
            }

            // 2. If same number exists in one array, exit
            if (unique_numbers.find(i) != unique_numbers.end())
            {
                return true;
            }

            unique_numbers.emplace(i);
        }
        // 3. This array is valid
        return false;
    };

    // 1. Check all the rows
    for (int r = 0; r < _grid_size; ++r)
    {
        if (sameNumberExistsInOneArray(getRow(r)))
            return false;
    }

    // 2. Check all the columns
    for (int c = 0; c < _grid_size; ++c)
    {
        if (sameNumberExistsInOneArray(getColumn(c)))
            return false;
    }

    // 3. Check all the boxes
    for (int r = 0; r < _box_size; ++r)
    {
        for (int c = 0; c < _box_size; ++c)
        {
            if (sameNumberExistsInOneArray(getBox(r * _box_size, c * _box_size)))
                return false;
        }
    }

    // 4. This is a valid sudoku
    return true;
}